src/deluge/gui/views/audio_clip_view.cpp

```diff
@@ -324,7 +325,46 @@ ActionResult AudioClipView::buttonAction(deluge::hid::Button b, bool on, bool in
 	}
 
 	else if (b == X_ENC) {
-		goto dontDeactivateMarker;
+		// removing time stretching by re-calculating clip length based on length of audio sample
+		if (Buttons::isShiftButtonPressed()) {
+			if (on && currentUIMode == UI_MODE_NONE) {
+				if (inCardRoutine) {
+					return ActionResult::REMIND_ME_OUTSIDE_CARD_ROUTINE;
+				}
+
+				AudioClip* audioClip = getCurrentAudioClip();
+				SamplePlaybackGuide guide = audioClip->guide;
+				SampleHolder* sampleHolder = (SampleHolder*)guide.audioFileHolder;
+				if (sampleHolder) {
+					Sample* sample = (Sample*)sampleHolder->audioFile;
+					if (sample) {
+						// get sample start position (green waveform marker)
+						int32_t loopStartSample =
+						    (uint32_t)(guide.getBytePosToStartPlayback(true) - sample->audioDataStartPosBytes)
+						    / (uint8_t)(sample->numChannels * sample->byteDepth);
+
+						// get sample end position (red waveform marker)
+						int32_t loopEndSample =
+						    (uint32_t)(guide.getBytePosToEndOrLoopPlayback() - sample->audioDataStartPosBytes)
+						    / (uint8_t)(sample->numChannels * sample->byteDepth);
+
+						uint32_t sampleLengthInSamples = std::abs(loopEndSample - loopStartSample);
+
+						// convert sample length in samples to ticks
+						float loopLength = (float)sampleLengthInSamples / playbackHandler.getTimePerInternalTickFloat();
```

Comment:
needs to be double - a 4gb audio file is longer than the largest int which can be stored in a float